[CORL-825] Allow sending moderation feedback to perspective

nick-funk opened the pull request:
If there is a `don't merge` label on this, it's because we haven't merged in `CORL-688` yet and this branch is based off of it.

## What does this PR do?

Adds a configuration option to send approved and rejected comment moderation actions to perspective to help improve the tracking model.

## How do I test this PR?

- Head to Admin > Config > Moderation > Toxic Comment Filter
- Enable/allow the `Allow Coral to send moderation actions to Google` option
- Create a comment with a commenter account
- Approve/Reject it
- It'll get sent to Perspective, feel free to check logs and debug into `src/core/server/services/perspective/index.ts` functions

wyattjoh committed:
feat: initial impl

nick-funk committed:
Create preliminary comment moderation slices

CORL-688

nick-funk committed:
Move slices logic into stacks

CORL-688

nick-funk committed:
Create user comment counts

CORL-688

nick-funk committed:
Create naive mutation that initializes user comment counts

CORL-688

nick-funk committed:
Use bulk updates in user counts migration

CORL-688

wyattjoh committed:
fix: review

wyattjoh committed:
fix: fixed issue with aggregation

nick-funk committed:
Migrate creating comment into stacks

CORL-688

nick-funk committed:
Migrate editing a comment to the stacks

CORL-688

nick-funk committed:
Break publishing comment status out of updateAllCounts

CORL-688

nick-funk committed:
Consolidate the tenant and common context together

CORL-688

nick-funk committed:
Allow sending of moderation decisions to Google

If enabled in the config, moderation decisions can
be sent to Google.

CORL-825

wyattjoh committed:
Merge branch 'master' into feature/CORL-688

wyattjoh committed:
Merge branch 'master' into feature/CORL-688

wyattjoh committed:
Merge branch 'master' into feature/CORL-688

wyattjoh committed:
review: removed variable scoping in favor of export

wyattjoh committed:
review: removed variable scoping in favor of export

wyattjoh committed:
revert: feb8e8196cd448f5cd24f1ca2eb0b91fe9bd43c7

wyattjoh committed:
revert: feb8e8196cd448f5cd24f1ca2eb0b91fe9bd43c7

wyattjoh commented on src/core/client/admin/routes/Configure/OnOffField.tsx:
```diff
@@ -15,6 +15,8 @@ interface Props {
   format?: (value: any, name: string) => any;
   parse?: (value: any, name: string) => any;
+  onTestId?: string;
+  offTestId?: string;
```
Perhaps something like this instead?

```ts
interface Props {
  // ...
  testIDs?: {
    on: string;
    off: string;
  }
}
```

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+  }
+  if (!config.sendFeedback) {
+    return;
+  }
```
Consolidate to:

```ts
if (!config.enabled || !config.key || !config.sendFeedback) {
  return;
}
```

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+    : TOXICITY_ENDPOINT_DEFAULT;
+  const apiKey = config.key;
+
+  const communityId = `Coral:${domain}`;
```
`domain != ROOT_URL`, this requires a fully resolved URL including scheme and port. See usage of `reconstructTenantURL` here:

https://github.com/coralproject/talk/blob/0f456330c5755c1faac3e7846204be3cad2bb73d/src/core/server/app/url.ts#L32-L48

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+
+  const communityId = `Coral:${domain}`;
+  const clientToken = `comment:${comment.id}`;
+  const latestRevision = comment.revisions[comment.revisions.length - 1];
```
The latest revision in this case should not be used. Because approve and reject requires you to specify the revision ID being approved/rejected, that revision ID should be used to extract the correct revision to send to perspective instead.

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+  const clientToken = `comment:${comment.id}`;
+  const latestRevision = comment.revisions[comment.revisions.length - 1];
+
+  const story = await findStory(mongo, comment.tenantID, {
```
Replace with `retrieveStory`, which already searches by ID.

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+  const commentStatus = computeStatus(status);
+  if (!commentStatus) {
+    return;
+  }
```
This can short circuit expensive database calls, move it right after the guard clauses referencing the `config`.

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+    id: comment.storyID,
+  });
+  if (!story) {
+    logger.error({ storyID: comment.storyID }, "could not find story");
```
```suggestion
    logger.warn({ storyID: comment.storyID }, "could not find story");
```

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+      clientToken,
+    };
+
+    const result = await send(endpoint, apiKey, "comments:suggestscore", body);
```
Add a `logger.debug` about sending the feedback to perspective.

wyattjoh commented on src/core/server/services/perspective/index.ts:
```diff
@@ -0,0 +1,146 @@
+  try {
+    const body = {
+      comment: {
+        text: latestRevision.body,
```
We only send the HTML stripped version of the body to perspective, ensure you strip the HTML prior to sending:

https://github.com/coralproject/talk/blob/0f456330c5755c1faac3e7846204be3cad2bb73d/src/core/server/services/comments/pipeline/pipeline.ts#L90

wyattjoh committed:
review: simplification of stacks implementation

This simplifies the stacks implementation to better reuse code related
to count management and event publishing. This can be used to great
effect with the upcomming events PR #2738.

wyattjoh committed:
review: simplification of stacks implementation

This simplifies the stacks implementation to better reuse code related
to count management and event publishing. This can be used to great
effect with the upcomming events PR #2738.

nick-funk committed:
fix: check if authorID is null before update user counts

CORL-688

nick-funk committed:
Merge remote-tracking branch 'origin/feature/CORL-688' into feature/CORL-688-context

# Conflicts:
#	src/core/server/stacks/helpers/updateAllCounts.ts

nick-funk committed:
Simplify OnOffField test id's

CORL-825

nick-funk committed:
Consolidate guard clauses

CORL-825

nick-funk committed:
Reconstruct tenant url's when sending perspective feedback

CORL-825

nick-funk committed:
Use the comment revision id when sending perspective feedback

CORL-825

nick-funk committed:
Use retrieveStory instead of findStory

CORL-825

nick-funk committed:
Clarify logging warning levels on perspective feedback

CORL-825

nick-funk committed:
Clean html tags from sent comment body

CORL-825

nick-funk committed:
Consolidate common/tenant context supporting files

CORL-688

nick-funk committed:
Merge remote-tracking branch 'origin/master' into feature/CORL-688-context

# Conflicts:
#	src/core/server/graph/mutators/Comments.ts
#	src/core/server/graph/tenant/mutators/Actions.ts
#	src/core/server/models/user/user.ts
#	src/core/server/services/comments/comments.ts
#	src/core/server/services/migrate/migrations/1575649180_user_comment_counts.ts
#	src/core/server/stacks/approveComment.ts
#	src/core/server/stacks/createComment.ts
#	src/core/server/stacks/editComment.ts
#	src/core/server/stacks/helpers/publishChanges.ts
#	src/core/server/stacks/rejectComment.ts

wyattjoh committed:
feat: renamed TenantContext -> GraphContext

nick-funk committed:
Fix include to match new graph context refactors

CORL-825
